File: distinct-subsequences-ii.cc
// Distinct Subsequences II
class Solution {
public:
  int distinctSubseqII(string a) {
    int mod = 1000000007, n = a.size(), s = 1;
    vector<int> bs(26, -1);
    for (int i = 0; i < n; i++) {
      int &b = bs[a[i]-'a'];
      b = exchange(s, (s*2L-(~b?b:0)+mod) % mod);
    }
    return s-1;
  }
};

File: distinct-subsequences-ii.rs
impl Solution {
    pub fn distinct_subseq_ii(s: String) -> i32 {
        const MOD: i64 = 1000000007;
        let mut dp = vec![0; 26];
        let mut sum = 0;
        for &c in s.as_bytes() {
            let x = dp[(c-b'a') as usize];
            dp[(c-b'a') as usize] = sum+1;
            sum += (sum+1-x) % MOD;
        }
        ((sum%MOD+MOD)%MOD) as i32
    }
}
